### DISCUSSION TITLE
Create Internal Repo from Template using REST API

### DISCUSSION THREAD

**Discussion by jrwhite17** [community/community#32817] [Apps, API and Webhooks]
Hello,

I am currently using GitHub AE. I'm wanting to write some code that generates internal repositories from an internal template (within an organization).

I've been following this documentation:
https://docs.github.com/en/github-ae@latest/rest/repos/repos#create-a-repository-using-a-template

When I use this REST API call: https://HOST.ghe.com/api/v3/repos/ORG/REPONAME/generate
{"owner":"ORG","name":"NEWREPONAME","description":"This is your first repository","include_all_branches":false, "private": false}

I receive: "message": "Public repositories not permitted on GitHub AE."

I am not seeing anything in the above documentation that describes how to generate an internal repository from a template. Is there a specific flag I need to pass in the body of the REST API call? Any information would be appreciated. Thanks!

**Comment by byrneh** [comment]
@jrwhite17  I see there is a body parameter for "vivibioity" for cerate a new repository but not one forr from template. 
Is your template visibility interal? If so what happens if you generrate with "private":true?

If no other option in a single API call you could create private and immediately folow up with with a 
PATCH /repos/{owner}/{repo} setting "visibility":"internal"

**Comment by pajones76** [comment]
Floating this back to the top of the list (hopefully?)

I'm having the same issue. I create a repository from a template using per [Create a repository using a template](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#create-a-repository-using-a-template). This takes no visibility setting. After that call I do a PATCH per [Update a repository](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#update-a-repository). This has a visibility setting which the documentation says can be only ````public```` or ````private````. 

When I try to pass ````internal```` anyway, it does not error and the text of the response after the call says ````'visibility': 'internal'````. However the created repository's visibility is not ````internal````. Since the template's visibility is ````internal````, I would at least hope that would carry through during the creation process.

Is this really intentional or is it an oversight? As a large enterprise organization we have lots of good reasons for setting an internal visibility.

**Comment by seshasandeep** [comment]
ditto

**Comment by chrisbarber86** [comment]
It would be great it this could be resolved. It means anyone enforcing policies to prevent the changing of visibility (usually to stop repos being made public by accident) - is then unable to create an internal repo from a template as they have to do it in two stages, create the repo as private and then change to internal - which is then blocked for most users. Why not allow visibility to be passed in this API rather than `private` as a boolean field.

**Comment by tylerablackham** [comment]
I am running into the same issue trying to use OpenTofu to generate repos from a template. Would love if you could pass the visibility in as internal for the API.
